src/main/scala/firrtl/Compiler.scala

```diff
@@ -246,6 +182,12 @@ abstract class Transform extends LazyLogging {
     state.annotations.collect { case a: LegacyAnnotation if a.transform == this.getClass => a }
   }
 
+  /** Executes before any transform's execute method
+    * @param state
+    * @return
+    */
+  def prepare(state: CircuitState): CircuitState = state
```

Comment:
I brought this up in a meeting ages ago but I want to touch base on it again.

I like the idea of `ResolvedAnnotationPaths` and being able to mix it in and get nice behavior like automatically running `resolvePathsOf` before execute. But it's not composable, you can't create something just like `ResolvedAnnotationPaths` that also overrides `prepare`, otherwise you will get a compiler error that you have to resolve by overriding `prepare` yourself.

I am, therefore, of the opinion that we should not yet expose `prepare` as a public API until we can figure out how to make it composable. I think if we make this a public API we're in for a load of hurt if/when we need to change it in the future.

Reply:
Do you have a suggestion for an alternative proposal? I agree with what you've said, but IMO the solution is schedulable transformations, which is a huge feature to add (and on my todo list) but way outside the scope of this PR. I could preemptively add a deprecation flag?

Reply:
Would making this an actual class that extends `Transform` and not a trait do it? At least then it's explicit that the user will have to override.

Reply:
@seldridge's suggestion would work, but in the composable analyses/prep like this world of my dreams, you could mixin multiple traits. Of course, traits have a potential ordering problem so it may just not be the right API.

In any case, I think prepare should just be `private[firrtl]` for now, then we can solve that problem later.

Reply:
Ok, will do.